lib/ex_plasma/transaction/type/payment_v1.ex

```diff
@@ -1,150 +1,249 @@
+
+  defp decode_outputs(outputs_rlp) do
+    {:ok, Enum.map(outputs_rlp, &Output.decode(&1))}
+  rescue
+    _ -> {:error, :malformed_outputs}
+  end
+
+  defp decode_metadata(metadata_rlp) when is_metadata(metadata_rlp), do: {:ok, metadata_rlp}
+  defp decode_metadata(_), do: {:error, :malformed_metadata}
+
+  defp decode_tx_data(@empty_tx_data), do: {:ok, @empty_tx_data}
+
+  defp decode_tx_data(tx_data_rlp) do
+    case RlpDecoder.parse_uint256(tx_data_rlp) do
+      {:ok, tx_data} -> {:ok, tx_data}
+      _ -> {:error, :malformed_tx_data}
+    end
+  end
+
+  @spec get_inputs(PaymentV1.t()) :: list(Output.t())
+  def get_inputs(%PaymentV1{} = transaction), do: transaction.inputs
+
+  @spec get_outputs(PaymentV1.t()) :: list(Output.t())
+  def get_outputs(%PaymentV1{} = transaction), do: transaction.outputs
+
+  @spec get_tx_type(PaymentV1.t()) :: pos_integer()
+  def get_tx_type(%PaymentV1{} = transaction), do: transaction.tx_type
+
   @doc """
   Validates the Transaction.
+  """
+  @spec validate(PaymentV1.t()) :: :ok | {:error, validation_error()}
+  def validate(%PaymentV1{} = transaction) do
+    with :ok <- validate_inputs(transaction.inputs),
+         :ok <- validate_outputs(transaction.outputs),
+         :ok <- validate_tx_data(transaction.tx_data),
+         :ok <- validate_metadata(transaction.metadata) do
+      :ok
+    end
+  end
 
-  ## Example
+  defp validate_inputs(inputs) do
+    with :ok <- validate_generic_output(inputs),
+         :ok <- validate_unique_inputs(inputs),
+         :ok <- validate_outputs_count(:inputs, inputs, 0) do
+      :ok
+    end
+  end
 
-  iex> txn = %{inputs: [%{output_data: [], output_id: %{blknum: 0, oindex: 0, position: 0, txindex: 0}, output_type: nil}], metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>, outputs: [%{output_data: %{amount: <<0, 0, 0, 0, 0, 0, 0, 1>>, output_guard: <<29, 246, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>, token: <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>}, output_id: nil, output_type: 1}], sigs: [], tx_data: <<0>>, tx_type: <<1>>}
-  iex> {:ok, ^txn} = ExPlasma.Transaction.Type.PaymentV1.validate(txn)
-  """
-  @impl Transaction
-  @spec validate(map()) :: validation_responses()
-  def validate(%{} = transaction) do
-    with {:ok, _inputs} <- do_validate_total(:inputs, transaction.inputs, 0),
-         {:ok, _outputs} <- do_validate_total(:outputs, transaction.outputs, 1) do
-      {:ok, transaction}
+  defp validate_outputs(outputs) do
+    with :ok <- validate_generic_output(outputs),
+         :ok <- validate_outputs_count(:outputs, outputs, 1),
+         :ok <- validate_outputs_type(outputs) do
+      :ok
     end
   end
 
-  defp do_validate_total(field, list, _min_limit) when length(list) > @output_limit do
+  defp validate_generic_output([output | rest]) do
+    with {:ok, _whatever} <- Output.validate(output), do: validate_generic_output(rest)
```

Comment:
single with issue statement

Reply:
I think childchain has a credo rule: https://github.com/rrrene/credo/pull/659/files that prevents this